Rework `.`-macro to support computed member exprs, remove `get`.

anko opened the pull request:
Details in a0fa525. Fixes #13. Breaks API, so would be a minor-version bump.

@tabatkins :+1:?

anko committed:
Rework `.`-macro to support computed member exprs

Fixes #13.

As pointed out by @wtabatkins in
<https://github.com/anko/eslisp/issues/13#issuecomment-142759925> , the
difference between the `.` and `get` macros has been confusing: The list
`(. a b)` has previously compiled to `a.b;`. This required a separate
form `(get a b)` to produce `a[b];`.

With these changes, `(. a b)` compiles to `a[b];`, and `a.b;` is
produced by `(. a "b")`.  This lets us discard the `get` macro, so the
`.` macro corresponds to a member expression with full generality.  This
keeps the language core minimal and easier to reason about.

This change does make the common use-case of long member expression
chains awkward to type if using core eslisp.  For example, what used to
be `(. Array prototype slice)` will need quotes around the properties:
`(. Array "prototype" "slice" "call")`.  However, if that is your
use-case, you should already be using macros (e.g.
[eslisp-propertify][1]) to sugar it.

If you preferred the old behaviour, you can still have it by writing the
appropriate macros.

[1]: https://github.com/anko/eslisp-propertify

anko commented:
Still needs appropriate changes to docs and doc-tests before merge, but this is what it would functionally look like.

anko commented:
Some relevant gitter chat about this [here](https://gitter.im/anko/eslisp?at=5603d95de85e8d337252a61a) and [here](https://gitter.im/anko/eslisp?at=56044d40131b784f781fbb23).

tabatkins commented on src/built-in-macros.ls:
```diff
@@ -244,20 +244,49 @@ contents =
+      # Ensure that a string literal is converted to an identifer and compiled
+      # into a non-computed member expression
+      if property-as-estree.type is \Literal
+      and typeof property-as-estree.value is \string
```
You need to ensure here that the value of the literal is valid in a dot expression.  It would suck if an author wrote `(. a "foo-bar")` (assuming it would desugar to the valid expressions `a["foo-bar"]`) and got an error instead.  (As currently written, they'd have to work around it with something like `(. a (+ "foo-bar"))` to force it down the "computed" path.)

Per the Gitter chat, while reserved keywords are now valid in dot expressions, it's probably also good to exclude them here, just to maximize compatibility with older browsers.

See https://mathiasbynens.be/notes/javascript-properties for details on what's valid or not.

anko replied:
The produced estree node is returned to [here](https://github.com/anko/eslisp/blob/a0fa525737f8f7f3c97c95220ae4c4a61fe390e2/src/ast.ls#L101-L109), which uses [esvalid](https://github.com/estools/esvalid) to check it for validity with [various criteria](https://github.com/estools/esvalid#validity-tests), including those.

tabatkins replied:
Right, that's the problem.  Typing `(. a "foo-bar")` (which _should_ be valid) will produce an AST error. There's no reason for that to error out, particularly since the `a.b` transformation (instead of `a["b"]`) is only done for perf anyway.  Instead, `(. a "foo-bar")` should be detected at this point, and instead desugar to `a["foo-bar"]` (in other words, follow the other branch).

anko committed:
In `.`, guard against invalid identifiers

Previously, `(. a "b-c")` would attempt to compile to `a.b-c`, where
`b-c` is an invalid identifier.

anko commented:
Having played with this for a while, I'm having second thoughts. I'll summarise.

Points in favour:
- Currently, `(. a "hi")` compiles to `a['hi'];`. One might reasonably expect that the literal `"hi"` can be replaced with an identifier containing the same value, like `(. a b)`, but that currently compiles to `a.b;`, which has a different meaning. This is confusing.
- This change means we would need 1 less macro, because `get` could go.

Points against:
- If `(. a "b")` compiles to `a.b;`, then how do you get `a["b"];`? (And vice-versa.) [The distinction matters](https://gitter.im/anko/eslisp?at=56044d40131b784f781fbb23). JavaScript has two constructs for member expressions; `.` for static and `[]` for computed, so we're losing generality.
- Static property access is by far more common than computed. Though [sugar](https://github.com/anko/eslisp-propertify) is available, writing `(. Array "prototype" "slice" "call")` to produce `Array.prototype.slice.call;` is quite grating compared to the current `(. Array property slice call)`.

anko commented:
After much further deliberation detailed in [the above comment](https://github.com/anko/eslisp/pull/14#issuecomment-143376495) and the original issue #13's [closing comments here](https://github.com/anko/eslisp/issues/13#issuecomment-143724145), I've concluded that these changes are unworkable.

If the described behaviour is of particular importance for some use-case, it could be implemented by a user macro.
